### PR TITLE
Reorder formatString expression check for JacksonEvent

### LINKED ISSUE #3357: [BUG] JSON parsing errors in raw processor
I launched a data-prepper pod on an EKS cluster to send traces from an otel-collector to an AWS managed OpenSearch.

The traces are being sent fine but the log on data-prepper keeps printing an error that seems to relate to JSON parsing on each record:
2023-09-18T19:48:33,795 [raw-pipeline-processor-worker-5-thread-1] DEBUG org.opensearch.dataprepper.pipeline.ProcessWorker - Pipeline Worker: Submitting 7 processed records to sinks
line 1:0 token recognition error at: 'spanId'
line 1:6 mismatched input '<EOF>' expecting {Function, Integer, Float, Boolean, 'null', JsonPointer, EscapedJsonPointer, VariableIdentifier, String, NOT, '-', '(', '{', OTHER}

**Steps to reproduce the behavior:**
1. Installed Istio as a service mesh.
2. Installed otel-collector otel/opentelemetry-collector:0.85.0 with a zipkin receiver
3. Configured Istio to send trace data to the otel-collector.
4. Otel-collector has a pipeline with zipkin as a receiver and data-prepper as exporter.
5. Data-prepper has a pipelines config as described in this project (https://github.com/opensearch-project/data-prepper/blob/main/examples/trace_analytics.yml) configured to the managed OpenSearch.

**Expected behavior**
No error logs or a log that allows me to fix the problem.

**Screenshots**
otel-collector config
Screenshot 2023-09-18 at 14 09 02

data-prepper config
Screenshot 2023-09-18 at 14 09 45

**Environment (please complete the following information):**
 - EKS. Kubernetes version: 1.27. 
 - Node OS: Amazon Linux 2. Kernel: 5.10.186-179.751.amzn2.x86_64
 - Data-prepper version: 2.4.0
 - Otel-collector version: 0.85.0
 - Istio version: 1.18.2

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description

Any call to JacksonEvent `formatString` method with an expression evaluator would call `isValidExpressionStatement`. A call to `isValidExpressionStatement` that was false, such as the configurations for normal keys (such as `${some_key}`) resulted in the ANTLR parser logging something like 

```
line 1:0 token recognition error at: 'spanId'
line 1:6 mismatched input '' expecting {Function, Integer, Float, Boolean, 'null', JsonPointer, EscapedJsonPointer, VariableIdentifier, String, NOT, '-', '(', '{', OTHER}
``` 
every time this occurs, which is very often when the format is a dynamic index that is evaluated for every Event. This causes cpu spikes and a thread freeze from timeouts for the logger writing to console. 

This mitigation includes swapping the order of the evaluation of the format expressions (`${}`), so that it is first evaluated whether or not the key is in the Event, and only if it is not do we evaluate whether the statement is a valid expression. The behavior should not change with this swap, but it will very heavily reduce the amount of logging that is seen from the ANTLR parser for this error. Eventually, we should try to keep this log from showing up completely when evaluating if an expression is valid.
 
### Issues Resolved
Resolves #3357 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by engechas — DISMISSED** [review]

**Review by oeyh — DISMISSED** [review]

**Review by kkondaka — DISMISSED** [review]

**Review by dlvenable — CHANGES_REQUESTED** [review]
Can you add a unit test to help prevent this from getting swapped again? Perhaps verify that if `val` is not `null` then there is no call to `evaluate()`.

**Review by engechas — APPROVED** [review]

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java [hunk 1]
```diff
@@ -567,8 +602,9 @@ public void testBuild_withFormatStringWithExpressionEvaluator() {
 
         final ExpressionEvaluator expressionEvaluator = mock(ExpressionEvaluator.class);
 
-        when(expressionEvaluator.isValidExpressionStatement("foo")).thenReturn(false);
+        verify(expressionEvaluator, times(0)).isValidExpressionStatement(eventKey);
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:605]: There is also a `never()` you can use instead of `times(0)`. It's a bit more readable, but should have the same result. Also, I'd probably put an `anyString()` in there instead of `eventKey`. When verifying that something is not called, I prefer to be as loose as possible.

**Review by dlvenable — APPROVED** [review]
Thanks for adding the unit test!
